homeassistant/components/zimi/config_flow.py

```diff
@@ -0,0 +1,114 @@
+"""Config flow for zcc integration."""
+
+from __future__ import annotations
+
+import logging
+import socket
+from typing import Any
+
+import voluptuous as vol
+from zcc import ControlPointDiscoveryService, ControlPointError
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_PORT
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.device_registry import format_mac
+
+from . import async_connect_to_controller
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_HOST): str,
+        vol.Optional(CONF_PORT, default=5003): int,
+        vol.Required(CONF_MAC): str,
+    }
+)
+
+
+class ZimiConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for zcc."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> config_entries.ConfigFlowResult:
+        """Handle the initial step."""
+
+        if user_input is not None:
+            data: dict[str, Any] = {"title": "ZIMI Controller"}
+            errors: dict[str, str] = {}
+            description_placeholders: dict[str, str] = {}
+
+            try:
+                if user_input[CONF_HOST] == "":
+                    try:
+                        description = await ControlPointDiscoveryService().discover()
+                    except ControlPointError as _:
+                        errors["base"] = "discovery_failure"
+                    data[CONF_HOST] = description.host
+                    data[CONF_PORT] = description.port
+                else:
+                    hostbyname = None
+                    data[CONF_HOST] = user_input[CONF_HOST]
+                    data[CONF_PORT] = user_input[CONF_PORT]
+                    try:
+                        hostbyname = socket.gethostbyname(data[CONF_HOST])
+                    except socket.gaierror as _:
+                        errors["base"] = "invalid_host"
+                    if hostbyname:
+                        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                        s.settimeout(10)
+                        try:
+                            s.connect((data[CONF_HOST], data[CONF_PORT]))
+                            s.close()
+                        except ConnectionRefusedError as _:
+                            errors["base"] = "connection_refused"
+                        except TimeoutError as _:
+                            errors["base"] = "timeout"
+                        except socket.gaierror as _:
+                            errors["base"] = "cannot_connect"
+
+                data[CONF_MAC] = format_mac(user_input[CONF_MAC])
+                if data[CONF_MAC] is user_input[CONF_MAC]:
+                    errors["base"] = "invalid_mac"
```

Comment:
what are you checking here?

Reply:
If the format_mac function is unable to correctly format the input string as a mac address it simply returns the same input string.    Therefore if the input and format_mac results are identical the format_mac failed and the mac is not valid.

Reply:
but what if I input my mac in the format that `format_mac` expects? It would block me then…

Reply:
Real-time review!!  

No, if format_mac is successful it returns a new string and the test on line 74 fails so the error is not set.    This would be cleaner if format_mac threw an Exception.

Reply:
Oh god, didn’t know `format_mac` was built like that and I hate it 😅